### PR TITLE
[core] Deprecate some syntax for ruleset references

### LINKED ISSUE #2352: [core] Deprecate <lang>-<ruleset> hyphen notation for ruleset references
> yes, we should just remove the example "java-basic". I don't think we have documented this notation anywhere.

_Originally posted by @adangel in https://github.com/_render_node/MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQxNTA1NDM5OnYy/pull_request_review_threads/discussion_

This notation is much less useful since we moved the rules into categories, and reduced the number of rulesets. Removing it for 7.0 should be considered.

### LINKED ISSUE #2532: [java]AvoidDecimalLiteralsInBigDecimalConstructor can not detect the case new BigDecimal(Expression)
**Affects PMD Version:** 
6.22.0

**Rule:**
AvoidDecimalLiteralsInBigDecimalConstructor

**Description:**
AvoidDecimalLiteralsInBigDecimalConstructor can not detect the case <code>new BigDecimal(Expression)</code>
```
//AllocationExpression
[ClassOrInterfaceType[@Image="BigDecimal"]]
[Arguments/ArgumentList/Expression/PrimaryExpression/PrimaryPrefix
    [
        Literal[(not(ends-with(@Image,'"'))) and contains(@Image,".")]
        or
        Name[ancestor::Block/BlockStatement/LocalVariableDeclaration
                [Type[PrimitiveType[@Image='double' or @Image='float']
                      or ReferenceType/ClassOrInterfaceType[@Image='Double' or @Image='Float']]]
                /VariableDeclarator/VariableDeclaratorId/@Image = @Image
            ]
        or
        Name[ancestor::MethodDeclaration/MethodDeclarator/FormalParameters/FormalParameter
                [Type[PrimitiveType[@Image='double' or @Image='float']
                      or ReferenceType/ClassOrInterfaceType[@Image='Double' or @Image='Float']]]
                /VariableDeclaratorId/@Image = @Image
            ]
    ]
]
```
This rule doesn't consider using Expression(for example: <code>foo * bar</code>) to create <code>BigDecimal</code>

**Code Sample demonstrating the issue:**
```
new BigDecimal(remaining * 60)
```

**Expected outcome:**
false-negative

**Running PMD through:**
CLI

### PR DESCRIPTION
**Pull request by oowekyala**
Some undocumented ruleset reference forms are deprecated:
- java-basic (resolves to rulesets/java/basic.xml), refs #2352
- 641 (resolves to rulesets/releases/641.xml)

Also using `<rule ref="rset1.xml,rset2.xml">` is deprecated. As before, only the first ruleset is considered (which was undocumented), but now the rest is not ignored silently.

I'll prepare the merge into PMD 7 and #3922 (which conflicts a lot).

## Related issues

- Refs #2612
- Fixes #2352

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/401849d0834f4d90941f09527d32a8d5/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java [hunk 1]
```diff
@@ -539,7 +546,7 @@ private void parseRuleSetReferenceNode(RuleSetBuilder ruleSetBuilder, Element ru
         // load the ruleset with minimum priority low, so that we get all rules, to be able to exclude any rule
         // minimum priority will be applied again, before constructing the final ruleset
         RuleSetFactory ruleSetFactory = toLoader().filterAbovePriority(RulePriority.LOW).warnDeprecated(false).toFactory();
-        RuleSet otherRuleSet = ruleSetFactory.createRuleSet(RuleSetReferenceId.parse(ref).get(0));
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java:542]: Ah... here we silently used only the first reference :)

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java [hunk 2]
```diff
@@ -743,6 +763,7 @@ && containsRule(ruleSetReferenceId, otherRuleSetReferenceId.getRuleName())) {
      * @return {@code true} if the ruleName exists
      */
     private boolean containsRule(RuleSetReferenceId ruleSetReferenceId, String ruleName) {
+        // unbelievable...
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java:766]: not helpful...

```suggestion
        // TODO: avoid reloading the ruleset once again
```

I guess, that's what you mean... right?

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetReferenceId.java [hunk 3]
```diff
@@ -85,12 +76,21 @@
 @Deprecated
 @InternalApi
 public class RuleSetReferenceId {
+
+    // todo this class has issues... What is even an "external" ruleset?
+    //  terminology and API should be clarified.
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetReferenceId.java:81]: Keep in mind at some point we marked it as InternalApi.
Clarification should happen anyway of course...

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetReferenceId.java [hunk 4]
```diff
@@ -177,8 +192,16 @@ public RuleSetReferenceId(final String id, final RuleSetReferenceId externalRule
...
             } else {
                 // resolve the ruleset name - it's maybe a built in ruleset
-                String builtinRuleSet = resolveBuiltInRuleset(tempRuleSetFileName);
+                String expandedRuleset = resolveDeprecatedBuiltInRulesetShorthand(tempRuleSetFileName);
+                String builtinRuleSet = expandedRuleset == null ? tempRuleSetFileName : expandedRuleset;
                 if (checkRulesetExists(builtinRuleSet)) {
+                    if (expandedRuleset != null && warnDeprecated) {
+                        LOG.warning(
+                            "Ruleset reference '" + tempRuleSetFileName + "' uses a deprecated form, use '"
+                            + builtinRuleSet + "' instead"
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetReferenceId.java:201]: :+1:

**Review by adangel — APPROVED** [review]
Thanks!

Let me know, if I should merge it or you want to do it...

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java (hunk 2 above)
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java:766]: Yeah sorry, I forgot about this... Will create an issue to track this later today

**Pull request merged**